File: problems-originals/1501-1600/1516_move-sub-tree-of-n-ary-tree/solution_parent_map.ts
function moveSubTree(root: Node | null, p: Node | null, q: Node | null): Node | null {
    // The judge hands over real nodes of the tree, so the case is never
    // hit in practice -- the signature keeps the optionals anyway.
    if (root === null || p === null || q === null) return root;
    // Pass one records every node's parent in a registry keyed by value
    // (the values are unique; the root has no entry); pass two probes p's
    // subtree for q. The surgery is the same three edits either way --
    // the registry is what answers the lookups.
    const parent = new Map<number, Node>();
    const stack: Node[] = [root];
    while (stack.length) {
        const node = stack.pop()!;
        for (const child of node.children) {
            parent.set(child.val, node);
            stack.push(child);
        }
    }
    let below = false;
    const probe: Node[] = [p];
    while (probe.length) {
        const node = probe.pop()!;
        if (node === q) {
            below = true;
            break;
        }
        probe.push(...node.children);
    }
    // p already hangs exactly where the move wants it: nothing to do.
    if (q.children.some((child) => child === p)) return root;
    if (below) {
        const qParent = parent.get(q.val)!;
        qParent.children.splice(qParent.children.indexOf(q), 1);
        if (!parent.has(p.val)) {
            // p is the root: q takes over.
            q.children.push(p);
            return q;
        }
        const holder = parent.get(p.val)!;
        holder.children[holder.children.indexOf(p)] = q;
        q.children.push(p);
        return root;
    }
    const pParent = parent.get(p.val)!;
    pParent.children.splice(pParent.children.indexOf(p), 1);
    q.children.push(p);
    return root;
}

File: problems/1501-1600/1516_transplanting-a-subtree/solution_parent_map.rs
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

impl Solution {
    pub fn transplant_subtree(
        root: Option<Rc<RefCell<SharedNode>>>,
        p: Option<Rc<RefCell<SharedNode>>>,
        q: Option<Rc<RefCell<SharedNode>>>,
    ) -> Option<Rc<RefCell<SharedNode>>> {
        let root = root?;
        let p = p?;
        let q = q?;
        // Pass one records every node's parent in a registry keyed by value
        // (the values are unique; the root has no entry). Identity stays
        // Rc::ptr_eq: p and q are real nodes of this tree.
        let mut parent: HashMap<i32, Rc<RefCell<SharedNode>>> = HashMap::new();
        let mut stack = vec![root.clone()];
        while let Some(node) = stack.pop() {
            for child in node.borrow().children.clone().into_iter().flatten() {
                let val = child.borrow().val;
                parent.insert(val, node.clone());
                stack.push(child);
            }
        }
        // Pass two probes p's subtree for q.
        let mut below = false;
        let mut probe = vec![p.clone()];
        while let Some(node) = probe.pop() {
            if Rc::ptr_eq(&node, &q) {
                below = true;
                break;
            }
            probe.extend(node.borrow().children.clone().into_iter().flatten());
        }
        // p already hangs exactly where the move wants it: nothing to do.
        let p_is_q_child = q.borrow().children.iter().flatten().any(|child| Rc::ptr_eq(child, &p));
        if p_is_q_child {
            return Some(root);
        }
        if below {
            let q_parent = parent[&q.borrow().val].clone();
            Self::detach(&q_parent, &q);
            match parent.get(&p.borrow().val) {
                None => {
                    // p is the root: q takes over.
                    q.borrow_mut().children.push(Some(p.clone()));
                    return Some(q);
                }
                Some(holder) => {
                    let slot = holder
                        .borrow()
                        .children
                        .iter()
                        .position(|child| child.as_ref().is_some_and(|c| Rc::ptr_eq(c, &p)))
                        .expect("p hangs from its parent");
                    holder.borrow_mut().children[slot] = Some(q.clone());
                    q.borrow_mut().children.push(Some(p.clone()));
                    return Some(root);
                }
            }
        }
        let p_parent = parent[&p.borrow().val].clone();
        Self::detach(&p_parent, &p);
        q.borrow_mut().children.push(Some(p.clone()));
        Some(root)
    }

    // detach removes node from parent's children list (identity is pointer
    // equality; the constraint guarantees the node hangs there).
    fn detach(parent: &Rc<RefCell<SharedNode>>, node: &Rc<RefCell<SharedNode>>) {
        parent
            .borrow_mut()
            .children
            .retain(|child| !child.as_ref().is_some_and(|c| Rc::ptr_eq(c, node)));
    }
}
